libfoundation/src/foundation-handler.cpp

```diff
@@ -301,6 +301,37 @@ static void __exec_closure(ffi_cif *cif, void *ret, void **args, void *user_data
         MCValueRelease(t_value_args[i]);
 }
 
+/* For some reason the current version / config of libffi on android 64-bit
+ * architectures fails to generate closure trampolines in memory with the
+ * exec bit set. To work-around this we set the exec bit of the memory page
+ * containing the function ptr directly. */
+#if defined(__ANDROID__) && (defined(__X86_64__) || defined(__ARM64__))
+
+#include <sys/mman.h>
+
+static bool ensure_block_is_executable(void *p_block, size_t p_size)
+{
+    /* Round the start of the block pointer down to the nearest page. */
+    byte_t *t_start_page = (byte_t *)(((uintptr_t)p_block) & ~4095);
+    
+    /* Round the end of the block pointer up to the nearest page. */
+    byte_t *t_finish_page = (byte_t *)((((uintptr_t)p_block + p_size) + 4096) & ~4095);
```

Comment:
I *think* this could protect an extra block if p_block+size is already aligned to block bounds - should be (uintptr_t)p_block + p_size + 4095

Reply:
Doh! Will fix